src/pre_processor.h

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file pre_processor.h
+ * @brief Interface for pre processing the input image before passing to
+ * `Inference::InferenceCore`
+ * 
+ * @copyright Copyright (C) 2021  Ashwin Maliampurakal
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef SRC_PRE_PROCESSOR_H_
+#define SRC_PRE_PROCESSOR_H_
+
+#include "inference_core.h"
+
+namespace PreProcessing {
+
+/**
+ * @brief A structure of the pre processed image
+ *
+ * After pre processing, the image will be resized and normalised before
+ * being passed to the `Inference::InferenceCore` as a `PreProcessedImage`
+ *
+ */
+typedef struct PreProcessedImage {
+  float* image;
+} PreProcessedImage;
+
+/**
+ * @brief PreProcess the input image before passing to `Inference::InferenceCore`
+ *
+ * This class contains some pre processing steps to allow an input image to be
+ * passed to the `Inference::InferenceCore`. The following operations are
+ * performed to process the raw data:
+ * - Resize the image to fit the dimensions of the model being used
+ * - Normalise the image pixels to the range [-1..1]
+ *
+ */
+class PreProcessor {
+ private:
+  size_t input_image_width;
+  size_t input_image_height;
+  size_t model_width;
+  size_t model_height;
+
+  /**
+   * @brief Resize the input image to the desired dimensions
+   *
+   * @param input_image Array of pixels of the input image
+   * @return The resized array of pixels of the input image
+   */
+  uint8_t* resize(uint8_t* input_image);
+
+  /**
+   * @brief Normalise the input image to the required interval [-1..1]
+   *
+   * @param resized_image Array of pixels of the resized input image  
+   * @return The normalised array of pixels of the input image
+   */
+  float* normalise(uint8_t* resized_image);
```

Comment:
Are you sure you want these to be "exposed" (even if they are `private`) in the API? They could be put in the .cpp file to keep things ambiguous at the interface level. I don't mind of they stay in the header, though. Just a point worth considering.

Reply:
I thought it might be nice to keep them in so it's quite clear what we actually mean by "pre processing" by just looking at this header file. We would then be able to add more in the future if we want to. I can take them out if you prefer though

Reply:
I would agreed with Ashwin here.

Reply:
Sounds good, just wanted to make sure.